### ISSUE TITLE
Error running example langevin_lj.yml

### ISSUE THREAD

**Issue by udayardahal** [mlund/faunus#305]
Dear @mlund,

I installed the latest langevin dynamics merged code and tried to run the example LJ fluid but stumbled with the following error. Could you please let me know if I did anything wrong during compilation or is it due to the bug in the code I installed?
The output looks as follows in verbose mode:

$ yason.py langevin_lj.yml | faunus -v 6
[faunus 13096] T: creating exclusion matrix 0×0 for 0 atoms within distance 0
[faunus 13096] T: creating exclusion matrix 1×0 for 1 atoms within distance 0
[faunus 13096] I: adding 100 active LJM molecules --> 25.8054 mol/l
[faunus 13096] I: plain pair-potential splined with 2 knots
[faunus 13096] D: Combination rule lorentz_berthelot in effect for the lennardjones potential.
[faunus 13096] T: Pair matrices for lennardjones sigma (1×1) and epsilon (1×1) created using 0 custom pairs.
[faunus 13096] D: Adding self-energy from coulomb/lennardjones to hamiltonian
[faunus 13096] E: vector::_M_fill_insert                     ]   2%  1.9/87.5s

**Comment by mlund** [comment]
Ping @rc83 @SHervoe-Hansen

**Comment by SHervoe-Hansen** [comment]
Hi @himalayanpride 
The problem is related to the generation of a high energy initial configuration configuration as `insertmolecules` does not check for any atomic overlap and with no minimization conducted before running the Langevin move the system will explode.

To accommodate this we have added a start,json which is an equilibrated configuration of the LJ spheres at 10 Kelvin obtained using Monte Carlo. You can run this example `yason.py langevin_lj.yml | faunus --state start.json`.

We would have to work on a minimization scheme to fix these overlaps, however at the time, I recommend you minimize with a low temperature Monte Carlo simulation followed by Langevin dynamics.

**Comment by SHervoe-Hansen** [comment]
@mlund One way we could sort of cheat our way out of this would be though the usage of Packmol, as the tolerance will make sure there will be no overlap.

**Comment by mlund** [comment]
> @mlund One way we could sort of cheat our way out of this would be though the usage of Packmol, as the tolerance will make sure there will be no overlap.

Yes, the current insertion schemes need updating according to issue #270.

**Comment by himalayanpride** [comment]
Thank you all for your input. I will use pre-equilibrated structure as input structure.

**Comment by himalayanpride** [comment]
How to implement Langevin dynamics along with the titration (rcmc)?  When I try to run the following script, I get the error message: 
`src/energy.cpp:651: virtual void Faunus::Energy::Bonded::force(std::vector<Eigen::Matrix<double, 3, 1> >&): Assertion absolute_index < forces.size()' failed.
`
Here is what my .yml script looks like.

```
#!/usr/bin/env yason.py

{% set aNaCl = 0.15 %}
{% set pH =  7 %}

random: {seed: default}
temperature: 300
geometry: {type: cuboid, length: 100.0 }
mcloop: {macro: 1, micro: 100}

atomlist:

    - na:   {q: 1.0,  sigma: 4.7, dp: 10, eps: 5.0}
    - cl:   {q: -1.0, sigma: 4.7, dp: 10, eps: 3.5}
    - AU:    {q: 0, sigma: 4.7, eps: 5.0, dp: 0}
    - C1:    {q: 0, sigma: 4.7, eps: 3.5, dp: 5}
    - C2:    {q: 0, sigma: 4.7, eps: 3.5, dp: 5}
    - C3:    {q: 0, sigma: 4.7, eps: 3.5, dp: 5}
    - Qd:    {q: 1, sigma: 4.7, eps: 5.0, dp: 5}
    - Qd0:   {q: 0, sigma: 4.7, eps: 5.0, dp: 5}
    - OW:   {q: 0, sigma: 4.7, eps: 4.0, mw: 24.0}
    - HW1:  {q: 0.46,  sigma: 0, eps: 0, mw: 24.0}
    - HW2:  {q: -0.46, sigma: 0, eps: 0, mw: 24.0}
    - H+:   { implicit: True, activity: {{ 10**(-pH) }} } 

moleculelist:

    - AUNP:
        structure: aunp_pos1.pqr
        bondlist:
           - harmonic: {index: [     0,    1], k: 25, req: 4.700}
           - harmonic: {index: [  1009, 1010], k: 1, req: 4.700}

    - water: 
        structure:
           - OW:  [1.060,  1.015,  0.979]
           - HW1: [0.964,  0.921,  0.941]
           - HW2: [0.975,  1.063,  1.080]
        bondlist:
           - harmonic: {index: [0,1], k: 100, req: 0.14}
           # all harmonic bonds
            .......
            .......
           - harmonic: {index: [0,2], k: 100, req: 0.14}
    - Na+: {atoms: [na], atomic: true, activity: {{aNaCl}} }
    - Cl-: {atoms: [cl], atomic: true, activity: {{aNaCl}} }

insertmolecules:

    - AUNP: {N: 1, positions: aunp_pos1.pqr, inactive: false}
    - Cl-: {N: 187, positions: cl_pos1.pqr, inative: false}
    - Na+: {N: 49, positions: na_pos1.pqr, inactive: false}
    - water: {N: 3764, rigid: true, inactive: false}
reactionlist:

    - Qd = Qd0 + H+: {pK: 10.23 }  # not electroneutral
    - = Na+ + Cl-: {} # activates grand canonical salt
energy:

    - isobaric: {P/mM: 113.2}
    - bonded: {}
    - nonbonded:
          default:
             - coulomb: {type: plain, epsr: 2.5}
             - lennardjones:
                mixing: LB
                custom:
                 - na cl: {eps: 4.0, sigma: 4.7}
                 - Qd cl: {eps: 4.0, sigma: 4.7}
                 - Qd0 cl: {eps: 4.0, sigma: 4.7}
                 - C1  Qd: {eps: 2.5, sigma: 4.7}
                 - C2  Qd: {eps: 2.5, sigma: 4.7}
                 - C3  Qd: {eps: 2.5, sigma: 4.7}
                 - C1 Qd0: {eps: 2.5, sigma: 4.7}
                 - C2 Qd0: {eps: 2.5, sigma: 4.7}
                 - C3 Qd0: {eps: 2.5, sigma: 4.7}
                 - C1  na: {eps: 2.5, sigma: 4.7}
                 - C2  na: {eps: 2.5, sigma: 4.7}
                 - C3  na: {eps: 2.5, sigma: 4.7}
                 - C1  cl: {eps: 2.5, sigma: 4.7}
                 - C2  cl: {eps: 2.5, sigma: 4.7}
                 - C3  cl: {eps: 2.5, sigma: 4.7}
                 - AU  na: {eps: 5.6, sigma: 4.7}
                 - AU  cl: {eps: 5.6, sigma: 4.7}
                 - AU  Qd: {eps: 5.6, sigma: 4.7}
                 - AU Qd0: {eps: 5.6, sigma: 4.7}
                 - AU  C1: {eps: 2.0, sigma: 4.7}
                 - AU  C2: {eps: 2.0, sigma: 4.7}
                 - AU  C3: {eps: 2.0, sigma: 4.7}
moves:

    - rcmc: {repeat: 1}
    - langevin_dynamics: # Langevin dynamics /w BAOAB splitting (Symmetric Langevin Velocity-Verlet)
        nsteps: 25
        integrator: {time_step: 0.005, friction: 5}
     #    - volume: {dV: 0.01, method: isotropic, repeat: 1}

analysis:

    - chargefluctuations: {nstep: 10, molecule: AUNP, pqrfile: AUNP_charge2.pqr} 
    - atomrdf: {file: rdf_ow.dat, nstep: 25, dr: 0.05, name1: OW, name2: OW, nskip: 200}
    - systemenergy: {file: energy2.dat, nstep: 100}
    - xtcfile: {file: traj2.xtc, nstep: 10}
    - savestate: {file: confout2.pqr}
    - savestate: {file: state2.json}
    - qrfile: {file: AUNP_qrtraj2.dat, molecule: AUNP,nstep: 10}
```

**Comment by mlund** [comment]
At the moment, shifting between MC and MD is to be considered experimental at best, and I don't think velocities are handled in the transition (ping @rc83 @SHervoe-Hansen ?). You could try a system without `rcmc` and equilibrate using MC, then run with MD. If this works as expected, we can use it as a test case for switching between MC/MD.

**Comment by himalayanpride** [comment]
Dear All,

What is the purpose of mcloop when doing langevin dynamics?

**Comment by mlund** [comment]
At the moment, langevin is run as if it is any other MC move, i.e. it is picked at random along with the other move. If picked, langevin will propagate the system using a user-defined number of steps (not to be confused with `micro` and `macro`) and will always be "accepted". Say you only have only move: langevin with nsteps=50; MC macro=10, MC micro=100. There will then be 10*100 calls to langevin, each call performing 50 LD steps. So in total 10*100*50 configurations. This behaviour can of course be discussed and we're happy to implement other schemes(?)

**Comment by himalayanpride** [comment]
Thank you @mlund  for the clarification. So, when we have multiple moves for example, if total transnational moves for water is N ; then per micro step we will have N water moves. So basically, we are already have many mc steps via moves even though we explicitly define micro and macro. There is no physical significance of micro and macro step other than the convenience for the user? Did I understand right?
Also there are a couple of issues i am facing:
 1) When I try to insert molecules from already equilibrated system ( here it is MARTINI CG MD simulation); FAUNUS throws the error saying that the (X) molecule is too large to insert and I had to significantly increase dimension of the box compared to pre-equilbrated system size. I am not sure why that's the case; as I am inserting all molecules from equilibrated structure via .pqr files. Does FAUNUS checks energy even when we are using external file to insert molecules?

2) When I try to use keepcharges: false (so that it uses charges from atomlist instead of .pqr file); it still is picking charges from .pqr file when there is mismatch in charges between the .pqr file and atomlist in the .yml file. Here is the syntax I used for.
```
moleculelist
    - water:
        structure:
           - OW:  [1.060,  1.015,  0.979]
           - HW1: [0.964,  0.921,  0.941]
           - HW2: [0.975,  1.063,  1.080]
        bondlist:
           - harmonic: {index: [0,1], k: 100, req: 0.14}
           - harmonic: {index: [0,2], k: 100, req: 0.14}
        keepcharges: false

```

**Comment by mlund** [comment]
Yes, macro and macro can be freely set, see [here](https://faunus.readthedocs.io/en/latest/_docs/topology.html#simulation-steps).

1. To insert an equilibrated system, you have to use a `.state` file describing the state of all molecules in the system. This can be created using `savestate`. Notice that when loading a state file (e.g. `faunus --state eq.state`) the input file should be the same as for the eq. run. That is, it must contain the same number of atoms and molecules, but can differ in terms of steps, moves and analysis.

2. `keepcharges` applies only when loading a structure for a specific molecules. Nothing seems to be loaded In the example above(?)

**Comment by himalayanpride** [comment]
For langevin for my nanoparticle case, I have tried with `state.pqr` file after I equilibrated with mc for around 1000 steps. The mc-equilibrium runs pretty smoothly but as soon as I input this final state file (`state.pqr`) with langevin turned on, I am getting error 
`src/energy.cpp:651: virtual void Faunus::Energy::Bonded::force(std::vector<Eigen::Matrix<double, 3, 1> >&): Assertion absolute_index < forces.size()' failed. `
I am not sure what structure is good for starting langevin in this particular case.  Please see this gdrive [link ](https://drive.google.com/drive/folders/1lf01_Mi-v0cwySBP4VankAvT41p5fruq?usp=sharing)for the files. Here I have used (...)1.pqr as starting insert structures and then ran MC simulation which gave me the output ` (state.json, confout.pqr, traj.xtc)`. After that  I turned on the `langevin_dynamics` but I got the error. 
To see how simple water system work, I tested for CG water alone and it runs fine. 

Also, If I want to freeze my gold atoms in position what do I need to do for langevin dynamics? I think this is also one of the issues for my system going forward. 

For the `keepcharges: false`; I only show a small segment of the input file.
I tried putting it in  insert molecules as below and also in moleculelist as mentioned earlier(that's is where the document says to put keepcharges=false) and neither seems to pick the charge from atomlist. I will recheck for a simple system since this is not a big issue for me at the moment as I can design my input file with correct charges.

```
[insertmolecules:]
    - AUNP: {N: 1, positions: aunp_pos1.pqr, inactive: false, keepcharges: false}
#    - Salt: {molarity: 0.15}
    - Cl-: {N: 187, positions: cl_pos1.pqr, inative: false, keepcharges: false}
    - Na+: {N: 49, positions: na_pos1.pqr, inactive: false, keepcharges: false}
    - water: {N: 3764, positions: water_pos1.pqr, rigid: true, inactive: false, keepcharges: false}
```

**Comment by mlund** [comment]
See attached input file with a reduced system without water molecules. Run first some steps without LD to relax overlaps, then activate it for hybrid MD/LD.
See if you can make your system run based on this setup, then add water. I also recommend that you install the `jsonschema` package and run `yason.py` from the faunus dir, i.e. `scripts/yason.py input.yml | faunus` which will trigger additional input control (I found a few invalid keys in the input file). 

The force failure is because forces are still unimplemented for the general `nonbonded` scheme. Instead use the hardcoded `nonbonded_coulomblj`, see input file below. Another issue is that the LD move brings molecular the mass centers out of sync; this is fixed in a patch you can activate like this: 

~~~ bash
git pull
git checkout forcemove-mass-center-sync
~~~

Inputfile:
[np_ld.yml.gz](https://github.com/mlund/faunus/files/4976099/np_ld.yml.gz)

**Comment by himalayanpride** [comment]
When I type this command: `git checkout forcemove-mass-center-sync`, I get the following error. 
error: pathspec 'forcemove-mass-center-sync' did not match any file(s) known to git. ( Sorry, I am new to git commands)

**Comment by mlund** [comment]
Sorry, my mistake, it should be `forcemove-masscenter-sync`.

**Comment by mlund** [comment]
@rc83: do you see a temporary way of adding OMP pragmas for efficiently evaluating nonbonded forces
in parallel, see code below? To avoid racing conditions for `forces[j]`, I thought of storing the pair-forces in a matrix first, but there could be other ways.
https://github.com/mlund/faunus/blob/6d48e0d104e9d859556c3391e89cb41285fc53a3/src/energy.h#L856-L866
(A serial run with added water will otherwise be very slow)

**Comment by himalayanpride** [comment]
Hi @mlund ,

I am able to successfully run the MD/MC for nanoparticle with ions. But I am having trouble adding water into the system. The error is `faunus/src/energy.cpp:651: virtual void Faunus::Energy::Bonded::force(std::vector<Eigen::Matrix<double, 3, 1> >&): Assertion `absolute_index < forces.size()' failed.`

This seems to be coming from bonded interaction as I turn off bonded energy term, the simulation runs fine. I tried to run purely water system and the same fate. If I try to add single water, even keeping bonded energy term is ok but adding more than one water molecule throws this error.  Since my nanoparticle also has bonded terms but I insert whole nanoparticle  as a single unit, I think the reason I didn't get error while simulating nanoparticle alone.

What could be the possible source of error here? See the attached file for water only system.

[water_ld.yml](https://drive.google.com/file/d/1HKCp5uxIr1HjBDef0EIurhVHa1tCLnwJ/view?usp=sharing)

**Comment by himalayanpride** [comment]
Dear @mlund,

I am able to resolve the issue of running langevin dynamics but now i am getting warning/error `Internal overflow compressing coordinates.` when running langevin dynamics.  Here is the simple rigid water model that I am testing.
```
geometry: {type: cuboid, length: 40}
mcloop: {macro: 10, micro: 1000}
temperature: 300.0
random: { seed: fixed }

energy:
#    - isobaric: {P/mM: 113.2}
    - bonded: {}
    - nonbonded_coulomblj:
        coulomb: {epsr: 1.3, type: plain}
        lennardjones: {mixing: LB}

atomlist:
    - OW:   {q: -2.00, sigma: 4.7, eps: 4.0, mw: 24.0, dp: 10}
    - HW1:  {q:  1.00, sigma: 0.0, eps: 0.0, mw: 24.0, dp: 10}
    - HW2:  {q:  1.00, sigma: 0.0, eps: 0.0, mw: 24.0, dp: 10}

moleculelist:
    - water:
        structure:
          - OW:  [0.34, 0.2, 0.09]
          - HW1: [0.07, -0.02, -1.06]
          - HW2: [-0.4, -0.17, 0.96]
#        bondlist:
#           - harmonic: {index: [0,1], k: 100, req: 1.2}
#           - harmonic: {index: [0,2], k: 100, req: 1.2}
#           - gromos_torsion: {index: [1,0,2], k: 4.2, aeq: 0}

insertmolecules:
    - water: {N: 300} 
moves:
#    - moltransrot: {molecule: water, dp: 10, dprot: 1, repeat: N}
#    - volume: {dV: 0.01, method: isotropic, repeat: 1}
    - langevin_dynamics: # Langevin dynamics /w BAOAB splitting (Symmetric Langevin Velocity-Verlet)
         nsteps: 10
         integrator: {time_step: 0.001, friction: 5}

analysis:
    - atomrdf: {file: rdf_ow_ld.dat, nstep: 25, dr: 0.05, name1: OW, name2: OW, nskip: 20}
     - savestate: {file: statewater_ld.json}
     - savestate: {file: watereq_ld.pqr}
     - xtcfile: {file: traj_ld.xtc, nstep: 10}
```

**Comment by mlund** [comment]
We will soon come back to the Langevin and fix some of the errors you're getting. One issue with the current setup is that the bonds an uncommented, meaning that hydrogen will quickly overlap with oxygen and cause an "implosion" (likely your overflow error). Also, checking you source code, no forces have been enabled for the gromos torsion energy, ping @rc83).

**Comment by himalayanpride** [comment]
Thank you @mlund and @rc83 . I guess you meant "bonds are commented?". Turning on the bonded energies doesn't allow me to simulate more than one water molecules, so I tried to constraint everything to make water-molecule rigid. I think that should already be sufficient to run the simulation as this is what exactly we do in regular martini simulation.
